browser/browsing_data/brave_clear_browsing_data_browsertest.cc

```diff
@@ -0,0 +1,383 @@
+                                   int remove_mask,
+                                   int origin_mask) override {
+    remove_data_call_count_++;
+
+    if (expected_remove_mask_ != -1)
+      EXPECT_EQ(expected_remove_mask_, remove_mask);
+    if (expected_origin_mask_ != -1)
+      EXPECT_EQ(expected_origin_mask_, origin_mask);
+  }
+
+ protected:
+  unsigned int browsers_count_ = 1u;
+  int remove_data_call_count_ = 0;
+  int expected_remove_data_call_count_ = 0;
+  int expected_remove_mask_ = -1;
+  int expected_origin_mask_ = -1;
+};
+
+IN_PROC_BROWSER_TEST_F(BraveClearDataOnExitTest, NoPrefsSet) {
+  // No set preferences to clear data.
+  SetExepectedRemoveDataCallCount(0);
+  // Tell the application to quit.
+  chrome::ExecuteCommand(browser(), IDC_EXIT);
+}
+
+IN_PROC_BROWSER_TEST_F(BraveClearDataOnExitTest, VerifyRemovalMasks) {
+  // Set all clear data on exit preferences and corresponding expected remove
+  // mask and origin flags.
+  SetClearAll(browser()->profile()->GetPrefs());
+
+  // Given those preferences the following removal mask is expected.
+  SetExpectedRemoveDataRemovalMasks(GetRemoveMaskAll(), GetOriginMaskAll());
+
+  // Expect a call to clear data.
+  SetExepectedRemoveDataCallCount(1);
+
+  // Tell the application to quit.
+  chrome::ExecuteCommand(browser(), IDC_EXIT);
+}
+
+class BraveClearDataOnExitTwoBrowsersTest : public BraveClearDataOnExitTest {
+ public:
+  BraveClearDataOnExitTwoBrowsersTest() : BraveClearDataOnExitTest() {
+    browsers_count_ = 2u;
+  }
+
+ protected:
+  // Open a new browser window with the provided |profile|.
+  Browser* NewBrowserWindow(Profile* profile) {
+    DCHECK(profile);
+    ui_test_utils::BrowserAddedObserver browser_added_observer;
+    chrome::NewEmptyWindow(profile);
+    Browser* browser = browser_added_observer.WaitForSingleNewBrowser();
+    DCHECK(browser);
+    content::WaitForLoadStopWithoutSuccessCheck(
+      browser->tab_strip_model()->GetActiveWebContents());
+    return browser;
+  }
+
+  // Open a new browser window with a guest session.
+  Browser* NewGuestBrowserWindow() {
+    ui_test_utils::BrowserAddedObserver browser_added_observer;
+    profiles::SwitchToGuestProfile(ProfileManager::CreateCallback());
+    Browser* browser = browser_added_observer.WaitForSingleNewBrowser();
+    DCHECK(browser);
+    // When a guest |browser| closes a BrowsingDataRemover will be created and
+    // executed. It needs a loaded TemplateUrlService or else it hangs on to a
+    // CallbackList::Subscription forever.
+    Profile* guest = g_browser_process->profile_manager()->GetProfileByPath(
+        ProfileManager::GetGuestProfilePath());
+    DCHECK(guest);
+    search_test_utils::WaitForTemplateURLServiceToLoad(
+        TemplateURLServiceFactory::GetForProfile(guest));
+    // Navigate to about:blank.
+    ui_test_utils::NavigateToURL(browser, GURL(url::kAboutBlankURL));
+    content::WaitForLoadStopWithoutSuccessCheck(
```

Comment:
I'm not sure if this call is needed, `ui_test_utils::NavigateToURL` is already a blocking call